File: aws-crt-cal/Hash.cs
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

using System;
using System.Runtime.InteropServices;

namespace Aws.Crt.Cal
{
    public class Hash
    {
        internal static class API
        {
            public delegate Handle aws_dotnet_sha1_new();
            public delegate Handle aws_dotnet_sha256_new();
            public delegate Handle aws_dotnet_md5_new();
            public delegate int aws_dotnet_hash_update(IntPtr hash,
                                                      [In, MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2, ArraySubType = UnmanagedType.U1)] byte[] buffer,
                                                       UInt32 buffer_length);
            public delegate void aws_dotnet_hash_digest(IntPtr hash, UInt32 truncate, 
                        [Out, MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 3, ArraySubType = UnmanagedType.U1)] byte[] buffer,
                                                       UInt32 buffer_length);
            public delegate void aws_dotnet_hash_destroy(IntPtr hash);
            public static aws_dotnet_sha1_new sha1_new = NativeAPI.Bind<aws_dotnet_sha1_new>();
            public static aws_dotnet_sha256_new sha256_new = NativeAPI.Bind<aws_dotnet_sha256_new>();
            public static aws_dotnet_md5_new md5_new = NativeAPI.Bind<aws_dotnet_md5_new>();
            public static aws_dotnet_hash_update update = NativeAPI.Bind<aws_dotnet_hash_update>();
            public static aws_dotnet_hash_digest digest = NativeAPI.Bind<aws_dotnet_hash_digest>();
            public static aws_dotnet_hash_destroy destroy = NativeAPI.Bind<aws_dotnet_hash_destroy>();

        }
        private static LibraryHandle library = new LibraryHandle();

        private Handle hash;
        private uint length;

        public class Handle : CRT.Handle
        {
            protected override bool ReleaseHandle()
            {
                API.destroy(handle);
                return true;
            }
        }

        private Hash(Handle hash, uint length)
        {
            this.hash = hash;
            this.length = length;
        }
        public static Hash sha1()
        {
            return new Hash(API.sha1_new(), 20);
        }
        public static Hash sha256()
        {
            return new Hash(API.sha256_new(), 32);
        }
        public static Hash md5()
        {
            return new Hash(API.md5_new(), 16);
        }

        public void update(byte[] buffer)
        {
            API.update(this.hash.DangerousGetHandle(), buffer, (uint)buffer.Length);
        }
        public byte[] digest(uint truncateTo = 0)
        {
            byte[] buffer = new byte[this.length];
            API.digest(this.hash.DangerousGetHandle(), truncateTo, buffer, this.length);
            return buffer;
        }
    }
}

File: native/src/hash.c
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

#include "aws/cal/hash.h"
#include "crt.h"
#include "exports.h"

AWS_DOTNET_API
struct aws_hash *aws_dotnet_sha1_new(void) {
    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    return aws_sha1_new(allocator);
}

AWS_DOTNET_API
struct aws_hash *aws_dotnet_sha256_new(void) {
    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    return aws_sha256_new(allocator);
}

AWS_DOTNET_API
struct aws_hash *aws_dotnet_md5_new(void) {
    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    return aws_md5_new(allocator);
}

AWS_DOTNET_API
int aws_dotnet_hash_update(struct aws_hash *hash, uint8_t *buffer, uint32_t buffer_size) {
    struct aws_byte_cursor buffer_cursor;
    AWS_ZERO_STRUCT(buffer_cursor);
    buffer_cursor.ptr = buffer;
    buffer_cursor.len = buffer_size;
    return aws_hash_update(hash, &buffer_cursor);
}

AWS_DOTNET_API
void aws_dotnet_hash_digest(struct aws_hash *hash, size_t truncate_to, uint8_t *buffer, uint32_t buffer_size) {
    (void)buffer_size;

    struct aws_byte_buf digest_buf = aws_byte_buf_from_array(buffer, hash->digest_size);
    digest_buf.len = 0;
    aws_hash_finalize(hash, &digest_buf, truncate_to);
}

AWS_DOTNET_API
void aws_dotnet_hash_destroy(struct aws_hash *hash) {
    aws_hash_destroy(hash);
}
